packages/sync/src/modules/class-full-sync-immediately.php

```diff
@@ -0,0 +1,405 @@
+		return $status;
+	}
+
+	/**
+	 * Get the range for content (posts and comments) to sync.
+	 *
+	 * @access private
+	 *
+	 * @return array Array of range (min ID, max ID, total items) for all content types.
+	 */
+	private function get_content_range() {
+		$range  = array();
+		$config = $this->get_status()['config'];
+		// Only when we are sending the whole range do we want to send also the range.
+		if ( true === isset( $config['posts'] ) && $config['posts'] ) {
+			$range['posts'] = $this->get_range( 'posts' );
+		}
+
+		if ( true === isset( $config['comments'] ) && $config['comments'] ) {
+			$range['comments'] = $this->get_range( 'comments' );
+		}
+
+		return $range;
+	}
+
+	/**
+	 * Get the range (min ID, max ID and total items) of items to sync.
+	 *
+	 * @access public
+	 *
+	 * @param string $type Type of sync item to get the range for.
+	 *
+	 * @return array Array of min ID, max ID and total items in the range.
+	 */
+	public function get_range( $type ) {
+		global $wpdb;
+		if ( ! in_array( $type, array( 'comments', 'posts' ), true ) ) {
+			return array();
+		}
+
+		switch ( $type ) {
+			case 'posts':
+				$table     = $wpdb->posts;
+				$id        = 'ID';
+				$where_sql = Settings::get_blacklisted_post_types_sql();
+
+				break;
+			case 'comments':
+				$table     = $wpdb->comments;
+				$id        = 'comment_ID';
+				$where_sql = Settings::get_comments_filter_sql();
+				break;
+		}
+
+		// TODO: Call $wpdb->prepare on the following query.
+		// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+		$results = $wpdb->get_results( "SELECT MAX({$id}) as max, MIN({$id}) as min, COUNT({$id}) as count FROM {$table} WHERE {$where_sql}" );
+		if ( isset( $results[0] ) ) {
+			return $results[0];
+		}
+
+		return array();
+	}
+
+	/**
+	 * Continue sending.
+	 *
+	 * @access public
+	 */
+	public function continue_enqueuing() {
+		$this->continue_sending();
+	} // Seconds.
+
+	/**
+	 * Continue sending.
+	 *
+	 * @access public
+	 */
+	public function continue_sending() {
+		if ( ! ( new Lock() )->attempt( self::LOCK_NAME ) || ! $this->is_started() || $this->get_status()['finished'] ) {
+			return;
+		}
+
+		$this->send();
+
+		( new Lock() )->remove( self::LOCK_NAME );
+	} // Seconds.
+
+	/**
+	 * Immediately send the next items to full sync.
+	 *
+	 * @access public
+	 */
+	public function send() {
+		$config = $this->get_status()['config'];
+
+		$max_duration = Settings::get_setting( 'full_sync_send_duration' );
+		$send_until   = microtime( true ) + $max_duration;
+
+		$progress = $this->get_status()['progress'];
+
+		/**
+		 * If a module exits early (e.g. because it ran out of full sync queue slots, or we ran out of request time)
+		 * then it should exit early
+		 */
+		foreach ( $this->get_remaining_modules_to_send() as $module ) {
+			$progress[ $module->name() ] = $module->send_full_sync_actions( $config[ $module->name() ], $send_until, $progress[ $module->name() ] );
+			if ( microtime( true ) >= $send_until ) {
```

Comment:
On top of the time limit checks, should we also be putting some memory limits to avoid any OOM errors?

Reply:
I will work on memory limits 👍 and another for number of requests to wpcom

Reply:
https://github.com/Automattic/jetpack/pull/13963/commits/6f831d1e96f027ea7aeb6d8389af9f7b32ed3bc4